File: src/main/java/com/clearcont/clearcontapp/model/Balancete.java
package com.clearcont.clearcontapp.model;

import com.clearcont.clearcontapp.helpers.DecimalFormatBR;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.proxy.HibernateProxy;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Entity
@Getter
@AllArgsConstructor
@NoArgsConstructor
public class Balancete {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @ManyToOne
    @Setter
    private Empresa empresa;
    @Setter
    private String nomeConta;
    @Setter
    private Integer numeroConta;
    private Double totalBalancete;
    @Setter
    private String classificacao;
    @Setter
    private String mes;
    @Setter
    private Integer ano;

    @OneToMany(mappedBy = "balancete", cascade = CascadeType.ALL, orphanRemoval = true)
    private @NotNull List<ComposicaoLancamentosContabeis> composicaoLancamentosContabeisList = new ArrayList<>();

    public void addComposicaoLancamentosContabeis(@NotNull ComposicaoLancamentosContabeis composicaoLancamentosContabeis) {
        this.composicaoLancamentosContabeisList.add(composicaoLancamentosContabeis);
        composicaoLancamentosContabeis.setBalancete(this);
    }

    public void removeComposicaoLancamentosContabeis(@NotNull ComposicaoLancamentosContabeis composicaoLancamentosContabeis) {
        this.composicaoLancamentosContabeisList.remove(composicaoLancamentosContabeis);
        composicaoLancamentosContabeis.setBalancete(null);
    }

    @Override
    public @NotNull String toString() {
        return "Balancete{" + "id=" + id + ", empresa=" + empresa + ", nomeConta='" + nomeConta + '\'' + ", numeroConta=" + numeroConta + ", totalBalancete=" + totalBalancete + ", classificacao='" + classificacao + '\'' + ", mes='" + mes + '\'' + ", ano=" + ano + '}';
    }

    public int contarPontos(@NotNull String texto) {
        int contador = 0;
        for (int i = 0; i < texto.length(); i++) {
            if (texto.charAt(i) == '.') {
                contador++;
            }
        }
        return contador;
    }

    public void setTotalBalancete(@NotNull String totalBalancete) {
        if (contarPontos(totalBalancete) < 1) this.totalBalancete = Double.valueOf(totalBalancete.replace(",", "."));
        else this.totalBalancete = Double.valueOf(totalBalancete.replaceFirst("\\.", "").replaceFirst(",", "."));
    }

    public @NotNull String getTotalBalancete() {
        return DecimalFormatBR.getDecimalFormat().format(totalBalancete);
    }

    public Double getDoubleTotalBalancete() {
        return totalBalancete;
    }

    @Override
    public final boolean equals(@Nullable Object o) {
        if (this == o) return true;
        if (o == null) return false;
        Class<?> oEffectiveClass = o instanceof HibernateProxy ? ((HibernateProxy) o).getHibernateLazyInitializer().getPersistentClass() : o.getClass();
        Class<?> thisEffectiveClass = this instanceof HibernateProxy ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass() : this.getClass();
        if (thisEffectiveClass != oEffectiveClass) return false;
        Balancete balancete = (Balancete) o;
        return getId() != null && Objects.equals(getId(), balancete.getId());
    }

    @Override
    public final int hashCode() {
        return this instanceof HibernateProxy ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass().hashCode() : getClass().hashCode();
    }
}

File: src/main/kotlin/br/com/clearcont/clearcontwebapp/models/Balancete.kt
package br.com.clearcont.clearcontwebapp.models

import br.com.clearcont.clearcontwebapp.utils.helpers.Period
import br.com.clearcont.clearcontwebapp.utils.helpers.formatCurrencyBR
import br.com.clearcont.clearcontwebapp.models.enums.StatusConciliacao
import br.com.clearcont.clearcontwebapp.models.enums.TipoConta
import br.com.clearcont.clearcontwebapp.models.enums.TypeCount
import jakarta.persistence.*
import jakarta.validation.constraints.NotNull
import java.time.LocalDate

@Entity
class Balancete(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,
    @ManyToOne
    var empresa: Empresa? = null,
    @Column(length = 155) var nomeConta: String = "",
    var numeroConta: Int = 0,
    private var totalBalancete: Double = 0.0,
    @Enumerated(EnumType.STRING)
    var classificacao: TypeCount = TypeCount.ATIVO,
    @Column(length = 30)
    var mes: String = Period.getPortugueseMonthName(LocalDate.now().month),
    val ano: Int = LocalDate.now().year,
    @OneToMany(
        mappedBy = "balancete",
        cascade = [CascadeType.REMOVE, CascadeType.DETACH, CascadeType.REFRESH],
        fetch = FetchType.EAGER,
        orphanRemoval = true
    )
    val lancamentosContabeisList: MutableList<ComposicaoLancamentosContabeis> = mutableListOf(),
    @Enumerated(EnumType.ORDINAL)
    @Column(length = 2)
    var status: StatusConciliacao = StatusConciliacao.OPEN,
    var tipo: TipoConta = TipoConta.INDEFINIDO
) {
    fun addComposicaoLancamentosContabeis(composicaoLancamentosContabeisLigth: ComposicaoLancamentosContabeis) {
        lancamentosContabeisList.add(composicaoLancamentosContabeisLigth)
        composicaoLancamentosContabeisLigth.balancete = this
    }

    fun removeComposicaoLancamentosContabeis(composicaoLancamentosContabeisLigth: ComposicaoLancamentosContabeis) {
        lancamentosContabeisList.remove(composicaoLancamentosContabeisLigth)
        composicaoLancamentosContabeisLigth.balancete = null
    }

    override fun toString(): String {
        return "Balancete{id=$id, empresa=$empresa, nomeConta='$nomeConta', numeroConta=$numeroConta, " +
                "totalBalancete=$totalBalancete, classificacao='$classificacao', mes='$mes', ano=$ano}"
    }

    fun getTotalBalancete(): String {
        return formatCurrencyBR(totalBalancete)
    }

    fun getTotalBalanceteDouble(): Double {
        return totalBalancete
    }
}
